File: ocaml/day2b.ml
open Base
open Stdio

let extract s =
  Caml.Scanf.sscanf s "%d-%d %c: %s" (fun min max c passw ->
      (min, max, c, passw))

let is_valid str char (pos1, pos2) =
  let c1 = str.[pos1 - 1] in
  let c2 = str.[pos2 - 1] in
  Poly.(c1 = char <> (c2 = char))

let rec count_valid_passwords ?(valid = 0) input_ch =
  let line = In_channel.input_line input_ch in
  match line with
  | None -> valid
  | Some l ->
    let pos1, pos2, c, passwd = extract l in
    if is_valid passwd c (pos1, pos2) then
      count_valid_passwords ~valid:(valid + 1) input_ch
    else count_valid_passwords ~valid input_ch

let solve filename = In_channel.with_file filename ~f:count_valid_passwords

let () = "../data/day2b.txt" |> solve |> printf "%d\n"

File: go/day2b.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	f, err := os.Open("../data/day2b.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	s := bufio.NewScanner(f)
	valid := 0
	var pos1, pos2 int
	var char byte
	var passwd string

	for s.Scan() {
		fmt.Sscanf(s.Text(), "%d-%d %c: %s", &pos1, &pos2, &char, &passwd)
		if (passwd[pos1-1] == char) != (passwd[pos2-1] == char) {
			valid++
		}
	}
	fmt.Println(valid)
}
